storage/src/main/scala/coop/rchain/storage/regex/PathRegex.scala

```diff
@@ -0,0 +1,239 @@
+package main.scala.coop.rchain.storage.regex
+
+import java.nio.charset.StandardCharsets
+import scala.annotation.tailrec
+
+private[regex] trait ParseOptions {
+  val delimiter: Char
+  val delimiters: Set[Char]
+}
+
+private[regex] trait RegexOptions extends ParseOptions {
+  val caseSensitive: Boolean
+  val strict: Boolean
+  val end: Boolean
+  val endsWith: Either[String, List[String]]
+}
+
+object PathRegexOptions {
+  val default       = PathRegexOptions()
+  val caseSensitive = PathRegexOptions(caseSensitive = true)
+  val strict        = PathRegexOptions(strict = true)
+  val nonEnd        = PathRegexOptions(end = false)
+}
+
+case class PathRegexOptions(
+    caseSensitive: Boolean = false,
+    strict: Boolean = false,
+    end: Boolean = true,
+    delimiter: Char = '/',
+    delimiters: Set[Char] = "./".toSet,
+    endsWith: Either[String, List[String]] = Right(Nil)
+) extends RegexOptions {}
+
+object PathRegex {
+  private[this] val defaultDelimiter = '/'
+
+  private[this] val defaultDelimiters = "./".toSet
+
+  private[this] val rxEscapeString = """([.+*?=^!:${}()\[\]|/\\])""".r
+
+  /**
+    * Escape a regular expression string.
+    */
+  def escapeString(str: String): String =
+    rxEscapeString.replaceAllIn(str, """\\$1""")
+
+  private[this] val utf8charSet = StandardCharsets.UTF_8
+
+  private[this] val uriAllowedChars =
+    (('a' to 'z') ++ ('A' to 'Z') ++ ('0' to '9') ++ "-_.!~*'()".toList).toSet
+
+  /**
+    * Accordingly to
+    * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent
+    * encodeURIComponent escapes all characters except: A-Z a-z 0-9 - _ . ! ~ * ' ( )
+    */
+  private[regex] def encodeUriComponent(str: String): String =
```

Comment:
Sure, before writing this implementation I've reviewed this topic and numerous others, including MDC spec. Finally I decided to use this implementation, since it's much more readable and straightforward than post-processing of URLEncoder output + has no dependencies (and therefore will not be broken if URLEncoder behavior changed).

Reply:
Cool.  I just wanted to make sure that you were aware of this.  

I agree that your solution is both readable and straightforward. 😄